<?xml version="1.0"?>
<robot name="arm">
  <!-- Base link of the robot -->
  <link name="world"/>

  <!-- Base link definition -->
  <link name="base_link">
    <visual>
      <geometry>
        <cylinder length="0.05" radius="0.2"/>
      </geometry>
      <material name="Black">
        <color rgba="0 0 0 1"/>
      </material>
      <origin rpy="0 0 0" xyz="0 0 0.025"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.05" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.025"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.025"/>
      <mass value="5.0"/>
      <inertia ixx="0.0135" ixy="0.0" ixz="0.0" iyy="0.0135" iyz="0.0" izz="0.05"/>
    </inertial>
  </link>

  <!-- Fixed joint connecting the base link to the world -->
  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Link 1 definition -->
  <link name="link_1">
    <visual>
      <geometry>
        <cylinder length="0.5" radius="0.08"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 0.8 1"/>
      </material>
      <origin rpy="0 0 0" xyz="0 0 0.25"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.5" radius="0.08"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.25"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.25"/>
      <mass value="5.0"/>
      <inertia ixx="0.107" ixy="0.0" ixz="0.0" iyy="0.107" iyz="0.0" izz="0.0125"/>
    </inertial>
  </link>

  <!-- Joint 1 definition (continuous) -->
  <joint name="joint_1" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="base_link"/>
    <child link="link_1"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 0.05"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Link 2 definition -->
  <link name="link_2">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.2"/>
      <mass value="2.0"/>
      <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
    </inertial>
    <visual>
      <geometry>
        <cylinder length="0.1" radius="0.08"/>
      </geometry>
      <material name="Red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.08"/>
      </geometry>
    </collision>
  </link>

  <!-- Joint 2 definition (continuous) -->
  <joint name="joint_2" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="link_1"/>
    <child link="link_2"/>
    <origin rpy="0 1.5708 0" xyz="0.0 -0.005 0.58"/>
    <limit lower="-0.25" upper="3.34" effort="10" velocity="0.5"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Link 3 definition -->
  <link name="link_3">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.2"/>
      <mass value="0.01"/>
      <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
    </inertial>
    <visual>
      <geometry>
        <cylinder length="0.4" radius="0.05"/>
      </geometry>
      <material name="blue">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.4" radius="0.05"/>
      </geometry>
    </collision>
  </link>

  <!-- Joint 3 definition (fixed) -->
  <joint name="joint_3" type="fixed">
    <parent link="link_2"/>
    <child link="link_3"/>
    <origin rpy="1.57 0 0" xyz="0.0 0.2 0"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Link 4 definition -->
  <link name="link_4">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.2"/>
      <mass value="0.01"/>
      <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
    </inertial>
    <visual>
      <geometry>
        <cylinder length="0.1" radius="0.06"/>
      </geometry>
      <material name="Red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.06"/>
      </geometry>
    </collision>
  </link>

  <!-- Joint 4 definition (continuous) -->
  <joint name="joint_4" type="continuous">
    <parent link="link_3"/>
    <child link="link_4"/>
    <origin rpy="1.57 0 0" xyz="0 0 -0.25"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.92" upper="1.92" effort="10" velocity="0.5"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Link 5 definition -->
  <link name="link_5">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.2"/>
      <mass value="0.01"/>
      <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
    </inertial>
    <visual>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
      <material name="yellow"> <!-- Fixed typo in material name -->
        <color rgba="0 1 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
      <dynamics damping="0.0" friction="0.0"/>
    </collision>
  </link>

  <!-- Joint 5 definition (fixed) -->
  <joint name="joint_5" type="fixed">
    <parent link="link_4"/>
    <child link="link_5"/>
    <origin rpy="1.57 0 0" xyz="0.0 -0.2 0"/>
    <dynamics damping="10" friction="1.0"/>
  </joint>

  <!-- Gazebo material definitions -->
  <gazebo reference="base_link">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="link_1">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="link_3">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="link_2">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="link_4">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="link_5">
    <material>Gazebo/White</material>
  </gazebo>

  <!-- Gazebo ROS2 Control Plugin -->
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <robot_sim_type>gazebo_ros2_control/GazeboSystem</robot_sim_type>
      <parameters>/home/kashyap/ros_ws/src/lab/config/control.yaml</parameters>
    </plugin>
  </gazebo>

  <!-- ROS2 Control System -->
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <!-- Joint definitions for control -->
    <joint name="joint_1">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <param name="initial_position">0.0</param>
    </joint>
    <joint name="joint_2">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <param name="initial_position">-1.57</param>
    </joint>
    <joint name="joint_4">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <param name="initial_position">0.0</param>
    </joint>
  </ros2_control>
</robot>
